### ISSUE TITLE
Undefined label for selected item when updating server side selectize input element

### ISSUE THREAD

**Issue by akaever** [rstudio/shiny#929]
There seems to be an issue with displayed labels when updating a selectizeInput element (server=TRUE, selected != NULL, and custom JS display function based on a data frame). The selected item is marked as `undefined` in the drop-down menu. See the following code:  

```
library(shiny)

runApp(list(
	ui = bootstrapPage(
		selectizeInput("selectizeTest", label="Selectize Test", choices=NULL)
	),
	server = function(input, output, session) {
		choices <- cbind(value = paste0(rownames(mtcars),".v"), label=paste0(rownames(mtcars),".l"), 
										name=paste0(rownames(mtcars),".n"), 
										html = paste0("<strong>", rownames(mtcars), "</strong>"), mtcars)

		updateSelectizeInput(session, "selectizeTest", choices=choices, 
												 selected="Cadillac Fleetwood.v", server = TRUE, 
												 options = list(render = I(
												 	'{
												 	option: function(item, escape) {
												 		return "<div>" + item.html + "</div>"; }
													}'
												 ))
		)
		# also for return "<div>" + escape(item.name) + "</div>" 
	}
))
```

This is most likely a timing problem. When updating the server-side selectize element some time after app initialization (e.g. using invalidateLater) with a new value, this item is displayed correctly according to the provided JS `option` function. In the first case, I guess, the data frame is not yet available and the `selected` parameter is interpreted as new value without label.

See also [http://shiny.rstudio.com/articles/selectize.html](http://shiny.rstudio.com/articles/selectize.html). It may be helpful to update this page (the JS function name needs to be `option` instead of `options`; it should be clear that the data frame has to contain a `value` and a `label` column;  the return value of the `option` function should always be wrapped in a `div` or similar tag, otherwise the items are not displayed correctly).

Thanks

**Comment by yihui** [comment]
Thanks for the report! I can reproduce it, and I'll see what I can do.

**Comment by jasdumas** [comment]
I've updated to the latest development version, and I'm trying to set the `selected = "num1"` from my data.frame and I'm still getting an undefined next to my value (aka item.label).

Should I be specifying the `selected = ` parameter differently from the changes you recently made @yihui ?

**Comment by yihui** [comment]
@jasdumas Please note my pull request #939 has not been merged yet, so you need to

```r
devtools::install_github('yihui/shiny@bugfix/selectize-selected')
```

if you want to test my fix (restart R before testing your app again). If that still does not work, could you provide a minimal reproducible example? Thanks!

**Comment by jasdumas** [comment]
That worked! Thanks - I will adjust accordingly once your pull request is merged

**Comment by yihui** [comment]
Excellent. Thanks for the confirmation!

**Comment by akaever** [comment]
The display of labels works, thanks! 

However, the order of items is somehow changed in the above example. Chrysler Imperial becomes the first element instead of Mazda RX4. This seems to happen when using server-side selectizeInputs and choices with more than 10 items. See the following example (which worked before the bug fix):

```
library(shiny)

runApp(list(
	ui = bootstrapPage(
		selectizeInput("selectizeTest", label="Selectize Test", choices=NULL)
	),
	server = function(input, output, session) {
		updateSelectizeInput(session, "selectizeTest", choices=as.character(1:11), 
												 selected="1", server = TRUE)
	}
))
```

**Comment by yihui** [comment]
@akaever Thanks! I just pushed a fix. Could you test `devtools::install_github('yihui/shiny@bugfix/selectize-selected')` again?

**Comment by akaever** [comment]
That works! However, there is now another issue: When trying to search for an item (typing backspace and search string), that has not been selected before, the selection always jumps to the item which was specified in the updateSelectizeInput call (server=TRUE). Once a particular item was selected, searching works fine (for this item). 

````
library(shiny)
 
runApp(list(
    ui = bootstrapPage(
        selectizeInput("selectizeTest", label="Selectize Test", choices=NULL)
    ),
    server = function(input, output, session) {
        updateSelectizeInput(session, "selectizeTest", choices=as.character(1:200), 
                                                 selected="2", server = TRUE)
    }
))
````

**Comment by yihui** [comment]
@akaever Ah too bad. Sorry about that. I just fixed it again. Please let me know if there are still further problems.

**Comment by akaever** [comment]
Thanks, that worked! There's just a side effect between the Shiny Server and updating the server side selectize elements. See [https://github.com/rstudio/shiny/issues/952](https://github.com/rstudio/shiny/issues/952)
